language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -129,436 +110,178 @@ impl AbstractState {
 
     /// destroys local@idx
-    /// call only if self.is_local_safe_to_destroy(idx) returns true
     pub fn destroy_local(&mut self, idx: LocalIndex) {
+        checked_verify!(self.is_local_safe_to_destroy(idx));
         let local = self.locals.remove(&idx).unwrap();
         match local {
             AbstractValue::Reference(nonce) => self.destroy_nonce(nonce),
-            AbstractValue::Value(kind, borrowed_nonces) => {
-                checked_verify!(kind == Kind::Unrestricted && borrowed_nonces.is_empty());
+            AbstractValue::Value(kind) => {
+                checked_verify!(kind == Kind::Unrestricted);
             }
         }
     }
 
-    /// nonce must be fresh
+    /// adds nonce to borrow graph
+    /// nonce must be freshly allocated
     pub fn add_nonce(&mut self, nonce: Nonce) {
-        self.partition.add_nonce(nonce);
+        self.borrow_graph.add_nonce(nonce);
     }
 
-    /// destroys nonce
-    /// borrows of nonce become borrows of any nonce' such that nonce borrows from nonce'
+    /// destroys nonce from borrow graph
     pub fn destroy_nonce(&mut self, nonce: Nonce) {
-        let mut nonce_set = BTreeSet::new();
-        let mut new_locals = BTreeMap::new();
-        let mut new_globals = BTreeMap::new();
-        let mut new_borrows = BTreeMap::new();
-
-        if let Some(borrow_info) = self.borrows.remove(&nonce) {
-            new_borrows = self.strong_propagate(nonce, &borrow_info);
-            match borrow_info {
-                BorrowInfo::BorrowedBy(borrowed_by) => {
-                    nonce_set = borrowed_by.clone();
-                }
-                BorrowInfo::FieldsBorrowedBy(fields_borrowed_by) => {
-                    nonce_set = Self::get_union_of_sets(&BTreeSet::new(), &fields_borrowed_by);
-                }
-            }
-        }
+        self.borrow_graph.destroy_nonce(nonce);
+    }
 
-        for (idx, value) in &self.locals {
-            if let AbstractValue::Value(kind, local_nonce_set) = value {
-                if local_nonce_set.contains(&nonce) {
-                    let mut local_nonce_set_restrict = local_nonce_set.clone();
-                    local_nonce_set_restrict.remove(&nonce);
-                    new_locals.insert(
-                        idx.clone(),
-                        AbstractValue::Value(
-                            *kind,
-                            local_nonce_set_restrict
-                                .union(&nonce_set)
-                                .cloned()
-                                .collect(),
-                        ),
-                    );
-                } else {
-                    new_locals.insert(
-                        idx.clone(),
-                        AbstractValue::Value(*kind, local_nonce_set.clone()),
-                    );
-                }
-            } else {
-                new_locals.insert(idx.clone(), value.clone());
-            }
-        }
+    /// returns all immediate borrows of nonce
```

Comment:
"return" is perhaps an unfortunate choice of words in the same sentence as "borrows". Perhaps "Gets"?

Reply:
Done